UnityProject/Assets/Scripts/Core/SecureStuff/SecureMapsSaver.cs

```diff
@@ -159,61 +654,74 @@ public static class SecureMapsSaver
 
 						IEnumerable list = null;
 
-						if (Field.FieldType.IsGenericType && (AMonoSet as IEnumerable) != null && Field.FieldType.GetGenericTypeDefinition() != typeof(Dictionary<,>) )
+						if (Field.FieldType.IsGenericType && typeof(IEnumerable).IsAssignableFrom(Field.FieldType) &&
+						    Field.FieldType.GetGenericTypeDefinition() != typeof(Dictionary<,>))
 						{
 							//&& Field.FieldType.GetGenericArguments()[0]
 
 							bool isScriptableObject = false;
 							bool IsComponent = false;
-							if (Field.FieldType.GetGenericArguments()[0].IsSubclassOf(typeof(GameObject)))
+
+							bool anyOfThem = false;
+
+							if (typeof(GameObject).IsAssignableFrom(Field.FieldType.GetGenericArguments()[0]))
 							{
 								IsComponent = false;
 								isScriptableObject = false;
+								anyOfThem = true;
 							}
-							else if (Field.FieldType.GetGenericArguments()[0].IsSubclassOf(typeof(Component)))
+							else if (typeof(Component).IsAssignableFrom(Field.FieldType.GetGenericArguments()[0]))
 							{
 								IsComponent = true;
 								isScriptableObject = false;
+								anyOfThem = true;
 							}
-							else if (Field.FieldType.GetGenericArguments()[0].IsSubclassOf(typeof(ScriptableObject)))
+							else if (typeof(ScriptableObject).IsAssignableFrom(
+								         Field.FieldType.GetGenericArguments()[0]) &&
+							         typeof(IHaveForeverID).IsAssignableFrom(Field.FieldType.GetGenericArguments()[0]))
 							{
 								IsComponent = false;
 								isScriptableObject = true;
+								anyOfThem = true;
 							}
 
-							var modified = IEnumeratorToList((AMonoSet as IEnumerable).GetEnumerator());
+							if (anyOfThem)
+							{
+								var modified = IEnumeratorToList((AMonoSet as IEnumerable).GetEnumerator());
 
-							List<object> original = new List<object>();
+								List<object> original = new List<object>();
 
 
-							if (APrefabDefault != null)
-							{
-								original = IEnumeratorToList((APrefabDefault as IEnumerable).GetEnumerator());
-							}
+								if (APrefabDefault != null)
+								{
+									original = IEnumeratorToList((APrefabDefault as IEnumerable).GetEnumerator());
+								}
 
 
-							for (int i = 0; i < Math.Max(original.Count, modified.Count); i++)
-							{
-								if (i < original.Count && i < modified.Count)
+								for (int i = 0; i < Math.Max(original.Count, modified.Count); i++)
 								{
-									if (CheckAreSame(original[i], modified[i]) == false)
+									if (i < original.Count && i < modified.Count)
+									{
+										if (CheckAreSame(original[i], modified[i]) == false)
+										{
+											SaveComponentGameObjectScriptableObject(IsComponent, isScriptableObject,
+												Prefix,
+												Field, i, modified, FieldDatas, UseInstance, IPopulateIDRelation);
+										}
+									}
```

Comment:
what is with you and giant functions and large nested if statements

pre-2020 yandere dev, is that you?

Reply:
yandere dev  doesn't use Switch statements xD